src/app/model/ChessEngine.ts

```diff
@@ -1,3 +1,48 @@
 import { ChessModel } from './ChessModel';
+import { Side, Square } from './Types';
+import { Piece } from './Piece';
+import { Player } from './Player';
+import { ChessBoard } from './ChessBoard';
+import { PieceWasMoved } from './PieceWasMoved';
+import { PieceWasCaptured } from './PieceWasCaptured';
 
-export class ChessEngine implements ChessModel {}
+export class ChessEngine implements ChessModel {
+  currentSide: Side = Side.BLACK;
+
+  constructor(readonly board: ChessBoard) {}
+
+  move(byPlayer: Player, squareFrom: Square, squareTo: Square): (PieceWasMoved | PieceWasCaptured)[] {
+    const chosenPiece = this.board.onPositionPiece(squareFrom) as Piece;
+    const possibleMoves = chosenPiece.possibleMoves(squareFrom, this.board);
+    const pieceOnSquare = this.board.onPositionPiece(squareTo);
+    const isOponent = pieceOnSquare && pieceOnSquare.side !== chosenPiece.side;
```

Comment:
Dobrze, zeby zmienne pojawialy sie tam gdzie sa uzywane :) Bo jak sa na gorze całej metody to nie mam jeszcze kontekstu.

Tak to mniej wiecej by mi wygladalo (plus, minus):

```
  move(byPlayer: Player, squareFrom: Square, squareTo: Square): (PieceWasMoved | PieceWasCaptured)[] {
    const chosenPiece = this.board.onPositionPiece(squareFrom) as Piece;
    if (byPlayer.side == this.currentSide) {
      throw new Error('Player can not move twice in a row.');
    }
    if (byPlayer.side !== chosenPiece.side) {
      throw new Error('Player can not move other players pieces.');
    }
    if (!this.canMoveOnSquare(chosenPiece.possibleMoves(squareFrom, this.board), squareTo)) {
      throw new Error('Piece can not move to given square.');
    }

    const pieceWasMoved: PieceWasMoved = { eventType: 'PieceWasMoved', piece: chosenPiece, from: squareFrom, to: squareTo };
    this.onPieceWasMoved(pieceWasMoved);
    
    const pieceWasCaptured = this.pieceWasCaptured(squareTo, chosenPiece);
    return pieceWasCaptured ? [pieceWasMoved, pieceWasCaptured] : [pieceWasMoved];
  }

  private pieceWasCaptured(squareTo: Square, chosenPiece: Piece): PieceWasCaptured | undefined {
    const pieceOnSquare = this.board.onPositionPiece(squareTo);
    const isOponent: boolean = pieceOnSquare !== undefined && pieceOnSquare.side !== chosenPiece.side;
    return isOponent && isPiece(pieceOnSquare)
      ? {
        eventType: 'PieceWasCaptured',
        piece: pieceOnSquare,
        onSquare: squareTo,
      } 
      : undefined;
  }
  ```